Add RODAS3.1 integration method to all Rosenbrock solvers

yantosca opened the pull request:
This PR seeks to add the RODAS3.1 integration method as an option (Mike Long et al 2025, in prep) to the Rosenbrock integrators.  Testing with GEOS-Chem has shown speedups of approximately 30% over the default RODAS3 method.

RODAS3 will still be the default Rosenbrock method, so there will be no differences when using the Rosenbrock integrators with "out-of-the-box" settings.  RODAS3.1 must be invoked with `ICNTRL(3) = 7`.

Manuscript: [Long et al. 2026](https://urldefense.proofpoint.com/v2/url?u=https-3A__earthandspacescience-2Dsubmit.agu.org_cgi-2Dbin_main.plex-3Fel-3DA4JI1HfIB4A5FrTx2F3A9ftdB0cZdqCxuS5ULUIvnvAJkwZ&d=DwMGAg&c=WO-RGvefibhHBZq3fL85hQ&r=XTJBopAzwSsAgEljIk6g1TdJEZxb_bVLEjVNpBxfBIw&m=QGxVzFDYnNDNoMNYsbO0qd03E7rCL-7y-ZAXLRrQpcY0cbYZ5vYPwLW9UifSe-jp&s=6tjrKTXRnzZCbAVtIRURqE2fvpOwr-nOK7keBlJLi2E&e=)

yantosca committed:
Add Rodas3.1 coefficients by Mike Long

int/rosenbrock.f90
int/rosenbrock_autoreduce.f90
- Added updated the Rosenbrock Rodas3 coefficients in routine
  Rodas3_1, which is called when ICNTRL(3)==0 or ICNTRL(3)==7.
- Trimmed trailing whitespace

CHANGELOG.md
- Updated accordingly

Signed-off-by: Bob Yantosca <yantosca@seas.harvard.edu>

yantosca committed:
Add Rodas3.1 coefficients by Mike Long

int/rosenbrock.f90
int/rosenbrock_autoreduce.f90
- Added updated the Rosenbrock Rodas3 coefficients in routine
  Rodas3_1, which is called when ICNTRL(3)==0 or ICNTRL(3)==7.
- Trimmed trailing whitespace

CHANGELOG.md
- Updated accordingly

Signed-off-by: Bob Yantosca <yantosca@seas.harvard.edu>

yantosca committed:
Add Rodas3.1 coefficients by Mike Long

int/rosenbrock.f90
int/rosenbrock_autoreduce.f90
- Added updated the Rosenbrock Rodas3 coefficients in routine
  Rodas3_1, which is called when ICNTRL(3)==7.
- Trimmed trailing whitespace

CHANGELOG.md
- Updated accordingly

Signed-off-by: Bob Yantosca <yantosca@seas.harvard.edu>

yantosca committed:
Add Rodas3.1 coefficients by Mike Long

int/rosenbrock.f90
int/rosenbrock_autoreduce.f90
- Added updated the Rosenbrock Rodas3 coefficients in routine
  Rodas3_1, which is called when ICNTRL(3)==7.
- Trimmed trailing whitespace

CHANGELOG.md
- Updated accordingly

Signed-off-by: Bob Yantosca <yantosca@seas.harvard.edu>

yantosca committed:
Update Rodas 3.1 coefficients again (as of 08 Apr 2025)

int/rosenbrock.F90
- Update Rodas 3.1 coefficients (@msl3v, 08 Apr 2025)
- Add comments to be consistent w/ rosenbrock_autoreduce.f90

int/rosenbrock_autoreduce.f90
- Update Rodas 3.1 coefficients (@msl3v, 08 Apr 2025)
- Fixed incorrect comment in ros_ErrorMsg, case -7
  (should be 0.1*H, not 10*H)

Signed-off-by: Bob Yantosca <yantosca@seas.harvard.edu>

yantosca committed:
Update Rodas 3.1 coefficients again (as of 08 Apr 2025)

int/rosenbrock.F90
- Update Rodas 3.1 coefficients (@msl3v, 08 Apr 2025)
- Add comments to be consistent w/ rosenbrock_autoreduce.f90

int/rosenbrock_autoreduce.f90
- Update Rodas 3.1 coefficients (@msl3v, 08 Apr 2025)
- Fixed incorrect comment in ros_ErrorMsg, case -7
  (should be 0.1*H, not 10*H)

Signed-off-by: Bob Yantosca <yantosca@seas.harvard.edu>

yantosca committed:
Add updated RODAS3.1 coeffs (22 May 2025), also add to C and Matlab

int/rosenbrock.c
int/rosenbrock.m
- Added Rodas3.1 routine with latest coeffs from Mike Long (22 May 2025)
- Trimmed trailing whitespace

int/rosenbrock.f90
int/rosenbrock_autoreduce.f90
- Updated Rodas3.1 routine with latest coeffs from Mike Long (22 May 2025)

CHANGELOG.md
- Updated accordingly

Signed-off-by: Bob Yantosca <yantosca@seas.harvard.edu>

yantosca committed:
Add updated RODAS3.1 coeffs (22 May 2025), also add to C and Matlab

int/rosenbrock.c
int/rosenbrock.m
- Added Rodas3.1 routine with latest coeffs from Mike Long (22 May 2025)
- Trimmed trailing whitespace

int/rosenbrock.f90
int/rosenbrock_autoreduce.f90
- Updated Rodas3.1 routine with latest coeffs from Mike Long (22 May 2025)

CHANGELOG.md
- Updated accordingly

Signed-off-by: Bob Yantosca <yantosca@seas.harvard.edu>

yantosca committed:
Add Rodas3.1 method to rosenbrock_adj.{c,f90} integrators

int/rosenbrock_adj.c
int/rosenbrock_adj.f90
- Add Rodas3_1 routine with updated coefficients by Mike Long (22 May 2025)

int/rosenbrock.c
- Fixed the definition of ros_Name in the Rodas3_1 routine

CHANGELOG.md
- Updated accordingly

Signed-off-by: Bob Yantosca <yantosca@seas.harvard.edu>

yantosca committed:
Add Rodas3.1 method to rosenbrock_adj.{c,f90} integrators

int/rosenbrock_adj.c
int/rosenbrock_adj.f90
- Add Rodas3_1 routine with updated coefficients by Mike Long (22 May 2025)

int/rosenbrock.c
- Fixed the definition of ros_Name in the Rodas3_1 routine

CHANGELOG.md
- Updated accordingly

Signed-off-by: Bob Yantosca <yantosca@seas.harvard.edu>

yantosca committed:
Fix compilation issues in rosenbrock_adj.c

int/rosenbrock_adj.c
- Added function prototype for Rodas3_1
- Changed "*ros_S" to "ros_S" in Rodas3_1 routine
- Changed "*ros_ELO" to ros_ELO" in Rodas3_1 routine

Signed-off-by: Bob Yantosca <yantosca@seas.harvard.edu>

yantosca committed:
Fix compilation issues in rosenbrock_adj.c

int/rosenbrock_adj.c
- Added function prototype for Rodas3_1
- Changed "*ros_S" to "ros_S" in Rodas3_1 routine
- Changed "*ros_ELO" to ros_ELO" in Rodas3_1 routine

Signed-off-by: Bob Yantosca <yantosca@seas.harvard.edu>

yantosca committed:
Add updated Rosenbrock coeffs by Mike Long (17 Jul 2025)

int/rosenbrock.c
int/rosenbrock.f90
int/rosenbrock.m
int/rosenbrock_adj.f90
int/rosenbrock_autoreduce.f90
int/rosenbrock_h211b_qssa.f90
int/rosenbrock_tlm.f90
- Update the Rodas3.1 coefficients to the set provided
  by Mike Long on 17 July 2025
- Update comments accordingly

Signed-off-by: Bob Yantosca <yantosca@seas.harvard.edu>

yantosca committed:
Add updated Rosenbrock coeffs by Mike Long (17 Jul 2025)

int/rosenbrock.c
int/rosenbrock.f90
int/rosenbrock.m
int/rosenbrock_adj.f90
int/rosenbrock_autoreduce.f90
int/rosenbrock_h211b_qssa.f90
int/rosenbrock_tlm.f90
- Update the Rodas3.1 coefficients to the set provided
  by Mike Long on 17 July 2025
- Update comments accordingly

Signed-off-by: Bob Yantosca <yantosca@seas.harvard.edu>

RolfSander committed:
rang -> rang3

RolfSander committed:
rang -> rang3

yantosca commented:
I have now rebased this PR atop PR #142.

yantosca commented:
Now rebased atop PR #143

yantosca commented:
@msl3v @RolfSander @obin1 I have rebased this PR atop KPP 3.3.1.

yantosca commented:
Hi @RolfSander @msl3v @obin1 @viral211, we would like to be able to use the RODAS3.1 option in GEOS-Chem 14.8.0 (to be released before June 2026).  Would it be OK if we merged this update as a "silent" feature (since it won't be the default option).

RolfSander committed:
Merge remote-tracking branch 'origin/feature/rodas3.1' into feature/rodas3.1

RolfSander committed:
rang -> rang3

RolfSander commented:
```
CASE (7)
CALL Rodas4
```
Shouldn't this be `CALL Rodas3.1` in `rosenbrock_adj.f90`?

yantosca committed:
Bug fix: Call RODAS3_1 for CASE(7) in rosenbrock_adj.f90

int/rosenbrock_adj.f90
- CASE(7) now calls RODAS3_1 instead of RODAS4, which was a typo.

Signed-off-by: Bob Yantosca <yantosca@seas.harvard.edu>

yantosca commented:
Thanks @RolfSander, this is now fixed in  f4642f23869eca3038deb696a82a2a10fc17254e

yantosca committed:
Updated CHANGELOG.md comment to include rosenbrock_tlm.f90

CHANGELOG.md
- Now also state that we have added RODAS3.1 to the
  rosenbrock_tlm.f90 integrator.

Signed-off-by: Bob Yantosca <yantosca@seas.harvard.edu>

RolfSander reviewed (approved):
Looks good now

yantosca commented:
Thanks @RolfSander.  I'll try to make a release soon, maybe with also a fix for the KPP runge_kutta.c integrator.
